blueocean-dashboard/src/main/js/components/LinkifiedText.jsx

```diff
@@ -0,0 +1,51 @@
+import React, { PropTypes } from 'react';
+import { Link } from 'react-router';
+
+const LinkifiedText = (props) => {
+    const { text, issues, textLink } = props;
+    const textWithIssues = [];
+
+    if (issues.length) {
+        let issuesIdString = '';
+        const issuesObj = {};
+
+        for (const issue of issues) {
+            issuesIdString += `${issue.id}|`;
+            issuesObj[issue.id] = issue.url;
+        }
+
+        if (issuesIdString) {
+            const issuesRegExpString = new RegExp(`(${issuesIdString.slice(0, -1)})`, 'gi');
+
+            for (let commitMsgPart of text.split(issuesRegExpString)) {
+                if (issuesObj[commitMsgPart]) {
+                    textWithIssues.push(<a href={issuesObj[commitMsgPart]} target="_blank">{commitMsgPart}</a>);
+                } else {
+                    if (commitMsgPart) {
+                        if (textLink) {
+                            textWithIssues.push((<Link to={textLink} className="unstyled-link" >{commitMsgPart}</Link>));
+                        } else {
+                            textWithIssues.push(commitMsgPart);
+                        }
+                    }
+                }
+            }
+        }
+
+        if (!textWithIssues.length) {
+            textWithIssues.push((<Link to={textLink} className="unstyled-link" >{text}</Link>));
+        }
+
+        return (<span>{textWithIssues}</span>);
+    }
+
+    return (textLink ? (<Link to={textLink} className="unstyled-link" >{text}</Link>) : (<span>{text}</span>));
+};
+
+LinkifiedText.propTypes = {
+    issues: PropTypes.object,
```

Comment:
This is an improvement, although it's still coupled to the `issues` structure which makes it difficult to reuse in other contexts - and linking to external systems is a pretty realistic future use case. Would you like me to take a quick stab at this or perhaps @michaelneale can weigh in on whether I'm being too picky :)

Reply:
No I can change the obj structure if needed, I just thought that it would be better to keep it as it comes from the backend because it was pretty clear and it was made especially for this use case (https://github.com/jenkinsci/blueocean-plugin/pull/1290), it just seems redundant to me to format it one way in the backend and then reformat it in the frontend before using it, maybe change the way it is sent from the backend?

Reply:
I am doing a poor job of explaining this, so let me try again:  suppose we have another place in the UI where we display a string of text and we want to replace certain portions of the text with hyperlinks. This could be Github PR's or anything else we dream up. The idea behind a generic `LinkifyText` component was that it would be generic and not coupled to JIRA, Github, or any other 3rd party service. So rather than having an `issues` prop, it would have a more generic prop of key/value pairs which indicate what a text fragment should be linked to. This component could be used in any context, JIRA, Github, whatever. 

In the specific case of JIRA, it would require some code outside of `LinkifyText` to transform the `issues` structure into the key/value pairs that `LinkifyText` expects. You could even create a `JiraIssueLinks` component which handles that data transform then uses the `LinkifyText`component internally. This kind of component composition is generally a favorable thing to strive for.

Please let me know if that makes sense. If not, I may create a quick branch tomorrow to demonstrate the idea.